Description: Allow Shift to interrupt hidden timeouts
 Upstream would like to consider this at more length. See
 http://lists.gnu.org/archive/html/grub-devel/2009-08/msg00718.html, and the
 rest of the thread for context.
Author: Colin Watson <cjwatson@ubuntu.com>
Forwarded: http://lists.gnu.org/archive/html/grub-devel/2009-08/msg00694.html
Last-Update: 2013-12-03

Index: b/grub-core/commands/sleep.c
===================================================================
--- a/grub-core/commands/sleep.c
+++ b/grub-core/commands/sleep.c
@@ -46,6 +46,31 @@
   grub_refresh ();
 }
 
+static int
+grub_check_keyboard (void)
+{
+  int mods = 0;
+  grub_term_input_t term;
+
+  if (grub_term_poll_usb)
+    grub_term_poll_usb ();
+
+  FOR_ACTIVE_TERM_INPUTS(term)
+  {
+    if (term->getkeystatus)
+      mods |= term->getkeystatus (term);
+  }
+
+  if (mods >= 0 &&
+      (mods & (GRUB_TERM_STATUS_LSHIFT | GRUB_TERM_STATUS_RSHIFT)) != 0)
+    return 1;
+
+  if (grub_getkey_noblock () == GRUB_TERM_ESC)
+    return 1;
+
+  return 0;
+}
+
 /* Based on grub_millisleep() from kern/generic/millisleep.c.  */
 static int
 grub_interruptible_millisleep (grub_uint32_t ms)
@@ -55,7 +80,7 @@
   start = grub_get_time_ms ();
 
   while (grub_get_time_ms () - start < ms)
-    if (grub_getkey_noblock () == GRUB_TERM_ESC)
+    if (grub_check_keyboard ())
       return 1;
 
   return 0;
Index: b/grub-core/normal/menu.c
===================================================================
--- a/grub-core/normal/menu.c
+++ b/grub-core/normal/menu.c
@@ -610,8 +610,27 @@
       saved_time = grub_get_time_ms ();
       while (1)
 	{
+	  int mods = 0;
+	  grub_term_input_t term;
 	  int key;
 
+	  if (grub_term_poll_usb)
+	    grub_term_poll_usb ();
+
+	  FOR_ACTIVE_TERM_INPUTS(term)
+	  {
+	    if (term->getkeystatus)
+	      mods |= term->getkeystatus (term);
+	  }
+
+	  if (mods >= 0 &&
+	      (mods & (GRUB_TERM_STATUS_LSHIFT
+		       | GRUB_TERM_STATUS_RSHIFT)) != 0)
+	    {
+	      timeout = -1;
+	      break;
+	    }
+
 	  key = grub_getkey_noblock ();
 	  if (key != GRUB_TERM_NO_KEY)
 	    {
